### ISSUE TITLE
Subobjects in proxyquired modules are not updating on re-proxyquing

### ISSUE THREAD

**Issue by Alexsey** [thlorenz/proxyquire#102]
<a href="https://github.com/Alexsey/proxyquire_bug">This project<a> contains working demonstration of the bug. Setup of a problem is
- we want to test module ```foo``` that requires module ```bar``` that contains object that contains function ```g```. We need to stub ```g```
- proxyquiring ```foo``` and stab ```bar``` with empty object ```fake```
- override ```g``` on ```fake``` (```g``` is not directly on ```bar```, so in is like ```fake.f.g = ...```)
- re-proxyquiring ```foo``` while stubbing ```bar``` with new empty object
- problem occurs - ```g``` is still overridden

For comparison in the root of ```bar``` there is also function ```h``` (not wrapped with ```f``` object) and it acts as expected

```noPreserveCache``` is active. So any changes on modules should be dropped on every new ```proxyquire```. But it is not

-

UPD: one of the possible solutions is to clean ```require.cache``` of ```bar``` before re-proxyquiring ```foo``` (like ```delete require.cache[require.resolve('./bar')]```). Not the thing I really want to do while writing simple tests...

-

UPD: this solution is even worse than it looks: if required ```foo``` require ```bar``` which requires ```baz``` then you need to clean cache for ```baz```. And you Definitely don't want to know about ```baz``` while you are testing ```foo```

-

May be I'm missing something and there is a better way to restore modules state between tests? For now it looks like much easier to stab all methods with ```sinon```'s ```sandbox```

**Comment by kilianc** [comment]
Not sure if I have the same problem but something similar is happening to me.

Consider the following code:

```js
const auth = proxyquire('lib/middleware/auth', {
  'lib/model': {
    AuthTokens: {
      refresh: (token) => (done) => done(null, token === 'auth-token')
    }
  }
})()
```

`AuthTokens.refresh` is permanently mocked even when `require('lib/middleware/auth').AuthTokens.refresh`, am I missing something or this is not how it's supposed to work?

**Comment by Alexsey** [comment]
Have you tried to use <a href="https://github.com/thlorenz/proxyquire#forcing-proxyquire-to-reload-modules">```noPreserveCache```</a>?

**Comment by bendrucker** [comment]
You definitely need noPreserveCache, we can reopen if that's not it

**Comment by kilianc** [comment]
@bendrucker it's the other way around, when I call `require` it returns the proxyquired mocked version. Not sure if my issue is separated.

**Comment by bendrucker** [comment]
Same thing. If you don't tell proxyquire not to use the require cache it will.

**Comment by Alexsey** [comment]
@bendrucker Look at example in initial message. There is ```noPreserveCace``` and it works bad

**Comment by Alexsey** [comment]
And it behave different for function that is directly in export object, and with on that is not directly on in

**Comment by kilianc** [comment]
@bendrucker ok, but I would consider changing this behavior anyway, I can't think of any case where you want proxyquire to globally mock a library and affect `require`, to me that's the whole point and reason why I am using it. Shouldn't it be the other way around?

If I am wrong I would personally rather export all the dependencies I have to mock in the test suite and have more control.

You may want to consider a `.restore()` function if the reason this is behaving like this purely a technicality.

**Comment by Alexsey** [comment]
@bendrucker could you please look at the initial problem and answer why this it not a bug, or reopen this issue?

**Comment by bendrucker** [comment]
Don't have time to look into it this week, will reopen in case anyone else wants to contribute

**Comment by kwhitaker** [comment]
I may have a related case to this, or it might be something that I'm misunderstanding about proxyquire.

In my case, I'm testing some methods normally, but then I need to stub those methods _later_. 

```
import * as methods from './some-module'
import proxyquire from 'proxyquire'
proxyquire.noPreserveCache()

... do some tests on `methods`
... later
const mock = proxyquire('./some-module', {
  someMethod: sinon.spy()
})
```

However, `mock` ends up being equal to `methods`, and therefore I can't spy on the internals.

**Comment by cbshakumar** [comment]
I think this may be related. I have been trying to understand if there was a way around this behavior.

```javascript
var proxyquire = require('proxyquire').noPreserveCache();
var expect = require('expect.js');
var TestModule = '../index.js';

describe('Test', function(){

  var dependencyMock;
  var test;
  before(function(){
    dependencyMock = function(){
      this.doStuff = function(callback){
        return callback(null, 'blah');
      };
    };
    var mocks = { './dependency.js': dependencyMock };
    var Test = proxyquire(TestModule, mocks);
    test = new Test();
  });

  it('does things', function(done){
    test.doSomething(function(err, data){
      expect(data).to.be('blah');
      done();
    });
  });

  describe('dependency returns banana', function(){
    before(function(){
      dependencyMock = function(){
        this.doStuff = function(callback){
          return callback(null, 'banana');
        };
      };

    });

    it('does banana things', function(done){
      test.doSomething(function(err, data){
        expect(data).to.be('banana');
        done();
      });
    });
  });
});
```

The test for banana fails because it still comes back with the original mocked data of 'blah'.

**Comment by Alexsey** [comment]
@cbshakumar Just before 'does banana things' test runs both of your ```before``` functions would be executed (```Mocha``` works that way) so ```doStuff``` would be reassigned twice. Probably ```Mocha``` execute firstly inner ```bofore``` and than outer ```before``` that is why in your test it returns 'blah' in stead of 'banana'. Log/debug your code to check the order of ```before```s execution

**Comment by Alexsey** [comment]
@kwhitaker you are trying to pass as second parameter of ```proxyquire``` call object with key that is the name of the method. ```proxyquire``` doesn't work that way: keys should be the names of packages that require module in first argument and values should be object with keys that are names of methods <b>in required</b> packages

**Comment by cbshakumar** [comment]
@Alexsey 
![screen shot 2016-05-04 at 4 18 04 pm](https://cloud.githubusercontent.com/assets/4872206/15027903/0636afb0-1214-11e6-90dc-9a5759155f3c.png)

before is run only once for each describe block it is in. This is not the beforeEach function from mocha.
But you can see from the repl/debug that it has already ran through the outer before statement by the time it has gotten to the inner one.

**Comment by kwhitaker** [comment]
@Alexsey thanks for clearing that up. I think I might need something like https://www.npmjs.com/package/rewire for this case.

**Comment by Alexsey** [comment]
@cbshakumar Yes, sorry, probably I confused them. Any way your problem is definitely in fact that you are reassigning your local ```dependencyMock``` variable with a new function while ```proxyquire``` continues to work with an old one. There is now way it JavaScript ```proxyquire``` can detect change of a target of a link

**Comment by cbshakumar** [comment]
@Alexsey right, I would imagine that just modifying the object instance itself, proxyquire should pick this up since it has the object reference still and I have no preserve cache set. How else can I get proxyquire to automatically pickup the modified object reference?

**Comment by Alexsey** [comment]
@cbshakumar ```proxyquire``` itself can't help you. But JS is flexible, so there are some tricks you can do. First I can imagine is to change ```mocks``` to
```
    var mocks = { './dependency.js': function () {
      return dependencyMock.apply(this, arguments)
    }};
```

**Comment by cbshakumar** [comment]
I was able to get tests passing by changing them to use an internal instance for the mock.

```javascript
var proxyquire = require('proxyquire').noPreserveCache();
var expect = require('expect.js');
var TestModule = '../index.js';

describe('Test', function(){

  var dependencyMock;
  var instance;
  var test;
  before(function(){
    console.log('outer');
    dependencyMock = function(){
      this.doStuff = function(callback){
        instance.doStuff(callback);
      };
    };
    instance = {
      doStuff: function(callback){
        return callback(null, 'blah');
      }
    };
    var mocks = { './dependency.js': dependencyMock };
    var Test = proxyquire(TestModule, mocks);
    test = new Test();
  });

  it('does things', function(done){
    test.doSomething(function(err, data){
      expect(data).to.be('blah');
      done();
    });
  });

  describe('dependency returns banana', function(){
    before(function(){
      instance.doStuff = function(callback){
        callback(null, 'banana');
      }
    });

    it('does banana things', function(done){
      test.doSomething(function(err, data){
        expect(data).to.be('banana');
        done();
      });
    });
  });
});

**Comment by Alexsey** [comment]
Yep, that is a similar way. I don't see how such helper could be a part of ```proxyquire```... but if you do - you should probably create a separate issue with suggestion as fare as we figure out that it is not about the bug from the initial post

**Comment by cbshakumar** [comment]
Sure

**Comment by bendrucker** [comment]
Thank you for the diligent effort in reproducing this @Alexsey 

I just dug in for a bit and was able to identify the issue. The stubs (child of a child) were still stuck in the require cache even though the child itself was deleted. I've made an addition to delete those as well. I've also incorporated a virtual identical unit test to your repro project and verified that the changes pass your tests.
